# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: editor/libeditor/InsertNodeTransaction.cpp
# Commit: ce2087ddab5c
# Full Hash: ce2087ddab5c2efcbce3b52da02b3363a37d6813
# Author: Masayuki Nakano <masayuki@d-toybox.com>
# Date: 2020-04-04 09:29:13
# Regressor Bug: 1619914
# File Overlap Count: 1
# Description:
#   Bug 1619914 - part 1: Make each transaction class grab their members with local variable before touching the DOM tree r=m_kato
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D69302
# ==============================================================================

diff -r 1c8115a9a684 -r ce2087ddab5c editor/libeditor/InsertNodeTransaction.cpp
--- a/editor/libeditor/InsertNodeTransaction.cpp	Fri Apr 03 10:21:33 2020 +0000
+++ b/editor/libeditor/InsertNodeTransaction.cpp	Fri Apr 03 08:30:37 2020 +0000
@@ -64,7 +64,7 @@
 InsertNodeTransaction::DoTransaction() {
   if (NS_WARN_IF(!mEditorBase) || NS_WARN_IF(!mContentToInsert) ||
       NS_WARN_IF(!mPointToInsert.IsSet())) {
-    return NS_ERROR_NOT_INITIALIZED;
+    return NS_ERROR_NOT_AVAILABLE;
   }
 
   if (!mPointToInsert.IsSetAndValid()) {
@@ -89,9 +89,9 @@
     }
   }
 
-  RefPtr<EditorBase> editorBase = mEditorBase;
-  nsCOMPtr<nsIContent> contentToInsert = mContentToInsert;
-  nsCOMPtr<nsINode> container = mPointToInsert.GetContainer();
+  OwningNonNull<EditorBase> editorBase = *mEditorBase;
+  OwningNonNull<nsIContent> contentToInsert = *mContentToInsert;
+  OwningNonNull<nsINode> container = *mPointToInsert.GetContainer();
   nsCOMPtr<nsIContent> refChild = mPointToInsert.GetChild();
   if (contentToInsert->IsElement()) {
     nsresult rv = editorBase->MarkElementDirty(
@@ -104,7 +104,7 @@
   }
 
   ErrorResult error;
-  container->InsertBefore(*contentToInsert, refChild, error);
+  container->InsertBefore(contentToInsert, refChild, error);
   if (error.Failed()) {
     NS_WARNING("nsINode::InsertBefore() failed");
     return error.StealNSResult();
@@ -132,10 +132,10 @@
   }
 
   // Place the selection just after the inserted element.
-  EditorRawDOMPoint afterInsertedNode(mContentToInsert);
-  DebugOnly<bool> advanced = afterInsertedNode.AdvanceOffset();
-  NS_WARNING_ASSERTION(advanced,
-                       "Failed to advance offset after the inserted node");
+  EditorRawDOMPoint afterInsertedNode(
+      EditorRawDOMPoint::After(contentToInsert));
+  NS_WARNING_ASSERTION(afterInsertedNode.IsSet(),
+                       "Failed to set after the inserted node");
   IgnoredErrorResult ignoredError;
   selection->Collapse(afterInsertedNode, ignoredError);
   NS_WARNING_ASSERTION(!ignoredError.Failed(),
@@ -156,8 +156,10 @@
   }
   // XXX If the inserted node has been moved to different container node or
   //     just removed from the DOM tree, this always fails.
+  OwningNonNull<nsINode> container = *mPointToInsert.GetContainer();
+  OwningNonNull<nsIContent> contentToInsert = *mContentToInsert;
   ErrorResult error;
-  mPointToInsert.GetContainer()->RemoveChild(*mContentToInsert, error);
+  container->RemoveChild(contentToInsert, error);
   NS_WARNING("nsINode::RemoveChild() failed");
   return error.StealNSResult();
 }